duopull/duopull.go

```diff
@@ -0,0 +1,698 @@
+		return nil, fmt.Errorf("type assertion failed on input event")
+	}
+
+	// Duo logging will store JSON data structures as strings in the log response,
+	// specifically with the event description field. Convert that into a
+	// structure here.
+	if x, ok := cv["event"]; ok {
+		if y, ok := x.(map[string]interface{}); ok {
+			if z, ok := y["description"]; ok {
+				if zs, ok := z.(string); ok {
+					ndesc := make(map[string]interface{})
+					err := json.Unmarshal([]byte(zs), &ndesc)
+					if err == nil { // on error, leave original intact
+						y["description"] = ndesc
+					}
+				}
+			}
+		}
+	}
+
+	err := flatten(cv, buf, []string{})
+	if err != nil {
+		return nil, err
+	}
+	l := log.New()
+	l.Formatter = &mozlogrus.MozLogFormatter{LoggerName: "duopull", Type: "app.log"}
+	bbuf := bytes.NewBuffer([]byte{})
+	l.Out = bbuf
+	l.WithFields(buf).Info("duopull event")
+	err = json.Unmarshal(bbuf.Bytes(), &ret)
+	return ret, err
+}
+
+func sendLogRequest(req *http.Request) ([]byte, error) {
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("API returned code %v with body %s", resp.StatusCode, b)
+	}
+	return b, nil
+}
+
+// logRequest makes a request for logs from the Duo API from mintime onwards, using the
+// specified API endpoint path for the request
+func logRequest(d *duoInterface, mintime int, path string) ([]emitEvent, error) {
+	mintimes := strconv.Itoa(mintime)
+
+	if debug == debugGCP {
+		// GCP debug, make an ad-hoc GET request to test outbound
+		// connectivity and then just return a test event
+		log.Info("making ad-hoc request")
+		resp, err := http.Get("https://www.mozilla.org")
+		if err != nil {
+			return nil, err
+		}
+		log.Infof("ad-hoc request returned status code %v\n", resp.StatusCode)
+		resp.Body.Close()
+		return []emitEvent{
+			{"/gcp/test", map[string]interface{}{
+				"gcp":       "test",
+				"timestamp": time.Now().Unix(),
+			}},
+		}, nil
+	}
+
+	req, err := http.NewRequest("GET", "https://"+d.apiHost+path, nil)
+	if err != nil {
+		return nil, err
+	}
+	q := req.URL.Query()
+	q.Add("mintime", mintimes)
+	req.URL.RawQuery = q.Encode()
+
+	authhdr, datehdr := d.getAuthHeader("GET", path, map[string]string{"mintime": mintimes})
+	req.Header.Set("Authorization", authhdr)
+	req.Header.Set("Date", datehdr)
+
+	b, err := sendLogRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	var l logRecords
+	err = json.Unmarshal(b, &l)
+	if err != nil {
+		return nil, err
+	}
+	if l.Stat != "OK" {
+		return nil, fmt.Errorf("%v invalid stat, got %v", path, l.Stat)
+	}
+	ret := make([]emitEvent, 0)
+	for _, v := range l.Response {
+		ret = append(ret, emitEvent{Path: path, Event: v})
+	}
+
+	return ret, nil
+}
+
+// logRequestV2
+func logRequestV2(d *duoInterface, mintime int, path string) ([]emitEvent, error) {
```

Comment:
It looks like there is a lot of duplicate code between this and `logRequest`, do we need both of them? No problem if so, just wondering if we still need to support the other version.

Reply:
Yeah, we do as there is no v2 for the admin and telephony logs.